File: signal-layer/src/api/kline.ts
import { api } from './client'
import { marketRequestTtl, recentMarketRequests } from './requestCache.ts'

export interface ApiKlineItem {
  open_time: number
  open: string
  high: string
  low: string
  close: string
  volume: string
  amount?: string
  turnover?: string
  turnover_rate?: string
  circulating_shares?: string
  adjustment_factor?: string
  adjustment_type?: string
  is_closed: boolean
}

export interface ApiKlineResponse {
  symbol: string
  timeframe: string
  data: ApiKlineItem[]
  from_time: number | null
  to_time: number | null
  count: number
  stale?: boolean
  refresh_failed?: boolean
  expected_time?: number | null
  status_message?: string | null
}

export interface FrontendKline {
  openTime: number
  open: number
  high: number
  low: number
  close: number
  volume: number
  amount?: number
  turnover?: number
  turnoverRate?: number
  circulatingShares?: number
  adjustmentFactor?: number
  adjustmentType?: string
  isClosed: boolean
}

export interface FrontendKlineResponse {
  symbol: string
  timeframe: string
  data: FrontendKline[]
  count: number
  toTime?: number | null
  stale: boolean
  refreshFailed: boolean
  expectedTime?: number | null
  statusMessage?: string | null
}

function toFrontend(item: ApiKlineItem): FrontendKline {
  return {
    openTime: item.open_time,
    open: parseFloat(item.open),
    high: parseFloat(item.high),
    low: parseFloat(item.low),
    close: parseFloat(item.close),
    volume: parseFloat(item.volume),
    ...(item.amount ? { amount: parseFloat(item.amount) } : {}),
    ...(item.turnover ? { turnover: parseFloat(item.turnover) } : {}),
    ...(item.turnover_rate ? { turnoverRate: parseFloat(item.turnover_rate) } : {}),
    ...(item.circulating_shares ? { circulatingShares: parseFloat(item.circulating_shares) } : {}),
    ...(item.adjustment_factor ? { adjustmentFactor: parseFloat(item.adjustment_factor) } : {}),
    ...(item.adjustment_type ? { adjustmentType: item.adjustment_type } : {}),
    isClosed: item.is_closed,
  }
}

export async function fetchKlines(
  symbol: string, timeframe: string, limit = 200,
): Promise<FrontendKlineResponse> {
  const key = `kline:${symbol}:${timeframe}:${limit}`
  return recentMarketRequests.run(key, marketRequestTtl(timeframe), async () => {
    console.log(`[SL:API] GET /klines/${symbol}`, { timeframe, limit })
    const raw = await api.get<ApiKlineResponse>(`/klines/${symbol}`, { timeframe, limit: String(limit) })
    console.log(`[SL:API] GET /klines/${symbol} -> OK ${raw.count} candles`)
    return {
      symbol: raw.symbol, timeframe: raw.timeframe, data: raw.data.map(toFrontend), count: raw.count,
      toTime: raw.to_time, stale: Boolean(raw.stale), refreshFailed: Boolean(raw.refresh_failed),
      expectedTime: raw.expected_time, statusMessage: raw.status_message,
    }
  })
}

/** 按时间范围读取 K 线。服务端会先查 DB，只从外部行情源补齐缺失区间。 */
export async function fetchKlineRange(
  symbol: string,
  timeframe: string,
  startTime: number,
  endTime: number,
): Promise<FrontendKlineResponse> {
  const raw = await api.get<ApiKlineResponse>(`/klines/${symbol}`, {
    timeframe,
    start_time: String(startTime),
    end_time: String(endTime),
    limit: '10000',
  })
  return {
    symbol: raw.symbol, timeframe: raw.timeframe, data: raw.data.map(toFrontend), count: raw.count,
    toTime: raw.to_time, stale: Boolean(raw.stale), refreshFailed: Boolean(raw.refresh_failed),
    expectedTime: raw.expected_time, statusMessage: raw.status_message,
  }
}

File: signal-layer-server/app/api/kline.py
from __future__ import annotations

import logging
from typing import Optional
from fastapi import APIRouter, Query, HTTPException, Depends
from app.api.deps import get_data_service
from app.api.security import require_permission
from app.schemas.kline import KlineResponse, KlineItem

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/klines", tags=["klines"])


@router.get("/{symbol}", response_model=KlineResponse)
async def get_klines(
    symbol: str,
    timeframe: str = Query(default="15m", description="Kline timeframe"),
    limit: int = Query(default=200, ge=1, le=10000),
    force_refresh: bool = Query(default=False),
    start_time: Optional[int] = Query(default=None, description="Range start, Unix milliseconds"),
    end_time: Optional[int] = Query(default=None, description="Range end, Unix milliseconds"),
):
    logger.info(
        f"GET /klines/{symbol} tf={timeframe} limit={limit} "
        f"start={start_time} end={end_time} force_refresh={force_refresh}"
    )
    if (start_time is None) != (end_time is None):
        raise HTTPException(status_code=422, detail="start_time and end_time must be provided together")
    if start_time is not None and end_time is not None and start_time > end_time:
        raise HTTPException(status_code=422, detail="start_time must be less than or equal to end_time")

    ds = get_data_service()
    try:
        result = await ds.fetch_klines(
            symbol=symbol,
            timeframe=timeframe,
            limit=limit,
            force_refresh=force_refresh,
            start_time=start_time,
            end_time=end_time,
        )
    except ValueError as e:
        raise HTTPException(status_code=422, detail=str(e)) from e
    except Exception as e:
        logger.error(f"GET /klines/{symbol} failed: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=str(e))

    items = [
        KlineItem(
            open_time=k["open_time"],
            open=str(k["open"]),
            high=str(k["high"]),
            low=str(k["low"]),
            close=str(k["close"]),
            volume=str(k["volume"]),
            amount=str(k.get("amount", 0)),
            turnover=str(k.get("turnover", 0)),
            turnover_rate=str(k["turnover_rate"]) if k.get("turnover_rate") is not None else None,
            circulating_shares=str(k["circulating_shares"]) if k.get("circulating_shares") is not None else None,
            adjustment_factor=str(k["adjustment_factor"]) if k.get("adjustment_factor") is not None else None,
            adjustment_type=k.get("adjustment_type"),
            is_closed=k.get("is_closed", True),
        )
        for k in result["data"]
    ]

    data_count = result["count"]
    logger.info(f"GET /klines/{symbol} -> {data_count} candles (cached={result.get('cached', False)})")
    return KlineResponse(
        symbol=result["symbol"],
        timeframe=result["timeframe"],
        data=items,
        from_time=result.get("from_time"),
        to_time=result.get("to_time"),
        count=data_count,
        cached=result.get("cached", False),
    )
